### PR TITLE
sql: change privilege validation errors to assertion errors

### PR DESCRIPTION
**Pull request by RichardJCai**
sql: change privilege validation errors to assertion errors

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/descpb/privilege.go [hunk 1]
```diff
@@ -355,15 +354,8 @@ func (p PrivilegeDescriptor) Validate(id ID, objectType privilege.ObjectType) er
 		}
 
 		if remaining := u.Privileges &^ allowedPrivilegesBits; remaining != 0 {
-			return fmt.Errorf("user %s must not have %s privileges on %s%s with ID=%d",
-				u.User(), privilege.ListFromBitField(remaining, privilege.Any), maybeSystem, objectType, id)
-		}
-		// Get all the privilege bits set on the descriptor even if they're not valid.
```
Comment (RichardJCai) [inline pkg/sql/catalog/descpb/privilege.go:361]: This was redundant with the above check.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65738#-:-Mae6yMe-yItOFJ-gL0V:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar and @RichardJCai)

___
*[pkg/sql/catalog/descpb/privilege.go, line 498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65738#-Mae6dS3BfMlWu-iqAbR:-Mae6dS3BfMlWu-iqAbS:b-76j181) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5b89d0f61ac477a314b93f9348da905943bbc0b/pkg/sql/catalog/descpb/privilege.go#L498)):*
> ```Go
> func maybeGetSystemString(id ID) (string, byte) {
> 	if IsReservedID(id) {
> 		return "system", ' '
> ```

`redact.SafeString`?

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65738)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @postamar)

___
*[pkg/sql/catalog/descpb/privilege.go, line 498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65738#-Mae6dS3BfMlWu-iqAbR:-MaeAAlMEWV2QC_Z7vv7:b-xvdn3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5b89d0f61ac477a314b93f9348da905943bbc0b/pkg/sql/catalog/descpb/privilege.go#L498)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`redact.SafeString`?
</blockquote></details>

Hm that seems to do it, thanks.

File: pkg/sql/grant_revoke.go [hunk 2]
```diff
@@ -209,6 +209,14 @@ func (n *changePrivilegesNode) startExec(params runParams) error {
 			n.changePrivilege(privileges, grantee)
 		}
 
+		// Ensure superusers have exactly the allowed privilege set.
+		// Postgres does not actually enforce this, instead of checking that
+		// superusers have all the privileges, Postgres allows superusers to
+		// bypass privilege checks.
+		if err := privileges.ValidateSuperuserPrivileges(descriptor.GetID(), n.grantOn); err != nil {
```
Comment (RichardJCai) [inline pkg/sql/grant_revoke.go:216]: Made a change so ValidateSuperuserPrivileges is a regular check that returns a non assertion error

However Validate calls ValidateSuperuserPrivileges and returns it as an assertion error. Reason being when we were granting/revoking, Validate was the only place we checked superuser privileges and if they're valid, outright replacing that check with assertion errors was incorrect.

**Comment by RichardJCai** [comment]
bors r=ajwerner

**Pull request merged**